frontends/systolic-lang/gen_post_op.py

```diff
@@ -0,0 +1,329 @@
+                valid_port
+                & (idx_port == cb.ExprBuilder(py_ast.ConstantPort(idx_width, max_idx)))
+            ) @ 1
+            this.computation_done = delay_reg.done @ 1
+
+
+def default_post_op(prog: cb.Builder, num_rows, num_cols, idx_width):
+    """
+    Adds a default post-op to `prog`.
+    This post-op does nothing except immediately write to memory.
+    """
+    comp = prog.component(name=DEFAULT_POST_OP)
+    add_post_op_params(comp, num_rows, idx_width)
+    for r in range(num_rows):
+        create_write_mem_groups(comp, r)
+    done_condition_groups(comp, num_rows, num_cols, idx_width)
+
+    comp.control = py_ast.StaticParComp(
+        [py_ast.Enable(WRITE_DONE_COND)]
+        # write to memory
+        + [py_ast.Enable(f"write_r{r}") for r in range(num_rows)]
+    )
+
+
+def leaky_relu_comp(prog: cb.Builder):
+    """
+    Creates a dynamic, non-pipelined, leaky relu component.
+    This is the component that actually performs the leaky relu computation on
+    a given output.
+    """
+    comp = prog.component(name="leaky_relu")
+    comp.input("value", BITWIDTH)
+    comp.input("index", BITWIDTH)
+    # Takes a memory and register (i.e., arguments that essentially act as ref cells)
+    add_write_mem_params(comp, OUT_MEM, BITWIDTH)
+    add_register_params(comp, "idx_reg", BITWIDTH)
+
+    this = comp.this()
+
+    addr0_port = cb.ExprBuilder.unwrap(this.port(OUT_MEM + "_addr0"))
+    write_data_port = cb.ExprBuilder.unwrap(this.port(OUT_MEM + "_write_data"))
+    write_en_port = cb.ExprBuilder.unwrap(this.port(OUT_MEM + "_write_en"))
+    write_done_port = this.port(OUT_MEM + "_done")
+
+    fp_mult = comp.fp_sop("fp_mult", "mult_pipe", BITWIDTH, INTWIDTH, FRACWIDTH)
+    lt = comp.fp_sop("val_lt", "lt", BITWIDTH, INTWIDTH, FRACWIDTH)
+    incr_idx = comp.add(BITWIDTH, "incr_idx")
+    write_mem = comp.wire("write_mem", 1)
+
+    with comp.continuous:
+        # gt holds whether this.value > 0
+        lt.left = this.value
+        lt.right = 0
+
+    with comp.group("do_relu") as g:
+        # Write_mem holds whether we should be writing to memory, which is when:
+        # a) multiplier is done, so we write fp_mult.out to mem
+        # b) this.value >=0 (i.e., !(this.value < 0)) so we write this.value to mem
+        write_mem.in_ = (fp_mult.done | ~lt.out) @ 1
+        # Trigger the multiplier when we're not writing to memory.
+        fp_mult.left = numeric_types.FixedPoint(
+            str(float_to_fixed_point(0.01, FRACWIDTH)), BITWIDTH, INTWIDTH, True
+        ).unsigned_integer()
+        fp_mult.right = this.value
+        fp_mult.go = ~(write_mem.out) @ 1
+
+        # Increment idx_reg during the cycle that we write to memory.
+        incr_idx.left = this.idx_reg_out
+        incr_idx.right = 1
+        this.idx_reg_in = write_mem.out @ incr_idx.out
+        this.idx_reg_write_en = write_mem.out @ 1
+
+        # Write to memory.
+        g.asgn(write_en_port, 1, write_mem.out)
```

Comment:
Why do we need to use `asgn` here instead of the `=` stuff?

Reply:
The port name depends on an input parameter `row_num`. So the port's name is `this.out_mem_{row_num}_write_en`.